### PR TITLE
Fix String creation to treat a modified UTF8 zero as ASCII

### PR DESCRIPTION
**Pull request by pshipton**
Updates j9gc_createJavaLangString().

Fixes https://github.com/eclipse-openj9/openj9/issues/19273

### CODE REVIEW

**Comment by pshipton** [comment]
jenkins test sanity alinux jdk8

**Comment by pshipton** [comment]
Opened https://github.com/eclipse-openj9/openj9/issues/19289 for the jdk8 test problem.

**Comment by pshipton** [comment]
jenkins test sanity alinux jdk8

File: runtime/gc_base/StringTable.cpp [hunk 1]
```diff
@@ -596,6 +597,12 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 	} else {
 		for (UDATA i = 0; i < length; ++i) {
 			if (data[i] > 0x7F) {
+				// check for 0 in modified UTF8
```
Comment (keithc-ca) [inline runtime/gc_base/StringTable.cpp:600]: Even though this is C++ code, C style comments should be used.

File: runtime/gc_base/StringTable.cpp [hunk 2]
```diff
@@ -623,7 +630,12 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 
 		if (isASCII) {
 			for (UDATA i = 0; i < length; ++i) {
-				hash = (hash << 5) - hash + data[i];
+				U_8 c = data[i];
+				if (0xC0 == c) {
```
Comment (keithc-ca) [inline runtime/gc_base/StringTable.cpp:634]: A comment would be helpful (somewhere) that explains why we don't need to check the second byte is 0x80 here and several places below.

File: runtime/gc_base/StringTable.cpp [hunk 3]
```diff
@@ -623,7 +630,12 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
...
+					c = 0;
+					i++;
```
Comment (keithc-ca) [inline runtime/gc_base/StringTable.cpp:636]: Please use `i += 1;` like line 603; same for lines 730, 745, 773 and 784.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by pshipton** [comment]
The alinux PR test https://openj9-jenkins.osuosl.org/job/Pipeline_Build_Test_JDK8_aarch64_linux/1761/

File: runtime/gc_base/StringTable.cpp (hunk 3 above)
Comment (pshipton) [reply runtime/gc_base/StringTable.cpp]: Updating again to tweak the comment wording.

**Comment by pshipton** [comment]
All the testing passed, the recent changes shouldn't affect that.

**Review by keithc-ca — APPROVED** [review]

**Comment by pshipton** [comment]
Actually, I'll update the test suite to include the case that fails without this change.

**Comment by pshipton** [comment]
Grinders on the updated test.

8: https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3464/ - passed
11: https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3465/ - passed
17: https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3466/ - passed

Grinder on jdk11 head stream
https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3468/ - failed test_lastIndexOf3 as expected

**Review by dmitripivkine — APPROVED** [review]

**Pull request merged**